File: frontend/src/types/log.ts
/**
 * Request Log Related Type Definitions
 * Corresponds to backend request_logs table
 */

/** Request Log Entity (For List View) */
export interface RequestLog {
  id: number;
  request_time: string;
  api_key_id?: number;
  api_key_name?: string;
  user_id?: string;
  requested_model?: string;
  target_model?: string;
  provider_id?: number;
  provider_name?: string;
  retry_count: number;
  first_byte_delay_ms?: number;
  total_time_ms?: number;
  input_tokens?: number;
  output_tokens?: number;
  total_cost?: number | null;
  input_cost?: number | null;
  output_cost?: number | null;
  response_status?: number;
  trace_id?: string;
  is_stream?: boolean;
}

/** Request Log Detail Entity (Includes full request/response) */
export interface RequestLogDetail extends RequestLog {
  detail_available?: boolean;
  request_headers?: Record<string, string>;  // Sanitized
  response_headers?: Record<string, string>; // Sanitized
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  request_body?: Record<string, any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  response_body?: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  usage_details?: Record<string, any>;
  error_info?: string;
  price_source?: 'SupplierOverride' | 'ModelFallback' | 'DefaultZero' | string | null;
  request_protocol?: string;
  supplier_protocol?: string;
  request_path?: string;
  request_url?: string;
  request_method?: string;
  upstream_url?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  converted_request_body?: Record<string, any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  upstream_response_body?: any;
}

export interface RetryLogResponse {
  response_status: number;
  response_body?: unknown;
  new_log_id?: number | null;
  trace_id?: string | null;
}

export interface ConvertedRequestResponse {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  converted_request_body?: Record<string, any> | null;
  upstream_url?: string | null;
  request_method?: string | null;
  supplier_protocol?: string | null;
}

export interface LogPlaygroundExecuteRequest {
  protocol: string;
  request_path?: string | null;
  request_headers?: Record<string, string>;
  request_body?: unknown;
}

export interface LogPlaygroundExecuteResponse {
  response_status: number;
  response_body?: unknown;
  trace_id?: string | null;
  provider_name?: string | null;
  target_model?: string | null;
  first_byte_delay_ms?: number | null;
  total_time_ms?: number | null;
}

/** Log Query Params */
export interface LogQueryParams {
  // Time range
  start_time?: string;
  end_time?: string;
  // Relative time range preset (server-side resolved). Ignored when start_time is set.
  timeline?: '1h' | '3h' | '6h' | '12h' | '24h' | '1w';

  // Client timezone offset minutes for stats bucketing (UTC to local)
  tz_offset_minutes?: number;

  // Trend bucketing hint for stats (hour/day)
  bucket?: 'minute' | 'hour' | 'day';
  bucket_minutes?: number;

  // Group by dimension for model stats
  group_by?: 'request_model' | 'provider_model';
  
  // Model filter
  requested_model?: string;
  target_model?: string;
  
  // Provider filter
  provider_id?: number;
  
  // Status code filter
  status_min?: number;
  status_max?: number;
  
  // Error filter
  has_error?: boolean;
  
  // API Key filter
  api_key_id?: number;
  api_key_name?: string;
  user_id?: string;
  
  // Retry count filter
  retry_count_min?: number;
  retry_count_max?: number;
  
  // Token range filter
  input_tokens_min?: number;
  input_tokens_max?: number;
  
  // Duration range filter
  total_time_min?: number;
  total_time_max?: number;
  
  // Pagination and Sorting
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface LogCostSummary {
  request_count: number;
  total_cost: number;
  input_cost: number;
  output_cost: number;
  input_tokens: number;
  output_tokens: number;
}

export interface LogCostTrendPoint {
  bucket: string;
  request_count: number;
  total_cost: number;
  input_cost: number;
  output_cost: number;
  input_tokens: number;
  output_tokens: number;
  error_count: number;
  success_count: number;
}

export interface LogCostByModel {
  requested_model: string;
  request_count: number;
  total_cost: number;
  input_tokens: number;
  output_tokens: number;
}

export interface LogCostStatsResponse {
  summary: LogCostSummary;
  trend: LogCostTrendPoint[];
  by_model: LogCostByModel[];
  by_model_tokens: LogCostByModel[];
}

File: backend/app/domain/log.py
"""
Request Log Domain Model

Defines Request Log related Data Transfer Objects (DTOs).
"""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.common.time import ensure_utc


class RequestLogBase(BaseModel):
    """Request Log Base Model"""
    
    # Request Time
    request_time: datetime = Field(..., description="Request Time")
    # API Key ID
    api_key_id: Optional[int] = Field(None, description="API Key ID")
    # API Key Name
    api_key_name: Optional[str] = Field(None, description="API Key Name")
    # User identifier from X-User-ID request header
    user_id: Optional[str] = Field(None, description="User ID")
    # Requested Model Name
    requested_model: Optional[str] = Field(None, description="Requested Model Name")
    # Target Model Name
    target_model: Optional[str] = Field(None, description="Target Model Name")
    # Provider ID
    provider_id: Optional[int] = Field(None, description="Provider ID")
    # Provider Name
    provider_name: Optional[str] = Field(None, description="Provider Name")
    # Whether the request has completed (False = still in progress)
    is_completed: bool = Field(True, description="Whether the request has completed")

    @field_validator("request_time", mode="after")
    @classmethod
    def _request_time_utc(cls, v: datetime) -> datetime:
        dt = ensure_utc(v)
        assert dt is not None
        return dt


class RequestLogCreate(RequestLogBase):
    """Create Request Log Model"""

    # Retry Count
    retry_count: int = Field(0, description="Retry Count")
    # First Byte Delay (ms)
    first_byte_delay_ms: Optional[int] = Field(None, description="First Byte Delay")
    # Total Time (ms)
    total_time_ms: Optional[int] = Field(None, description="Total Time")
    # Input Token Count
    input_tokens: Optional[int] = Field(None, description="Input Token Count")
    # Output Token Count
    output_tokens: Optional[int] = Field(None, description="Output Token Count")
    # Cost fields (USD, 4 decimals)
    total_cost: Optional[float] = Field(None, description="Total cost ($)")
    input_cost: Optional[float] = Field(None, description="Input cost ($)")
    output_cost: Optional[float] = Field(None, description="Output cost ($)")
    # Cached cost fields
    cached_input_cost: Optional[float] = Field(None, description="Cached input cost ($)")
    cached_output_cost: Optional[float] = Field(None, description="Cached output cost ($)")
    # Price source: SupplierOverride / ModelFallback / DefaultZero
    price_source: Optional[str] = Field(None, description="Price source")
    # Request Headers (Sanitized)
    request_headers: Optional[dict[str, Any]] = Field(None, description="Request Headers")
    # Response Headers (Sanitized)
    response_headers: Optional[dict[str, Any]] = Field(None, description="Response Headers")
    # Request Body
    request_body: Optional[dict[str, Any]] = Field(None, description="Request Body")
    # Response Status Code
    response_status: Optional[int] = Field(None, description="Response Status Code")
    # Response Body
    response_body: Optional[str] = Field(None, description="Response Body")
    # Usage Details (normalized)
    usage_details: Optional[dict[str, Any]] = Field(None, description="Usage Details")
    # Error Info
    error_info: Optional[str] = Field(None, description="Error Info")
    # Matched Provider Count
    matched_provider_count: Optional[int] = Field(None, description="Matched Provider Count")
    # Trace ID
    trace_id: Optional[str] = Field(None, description="Trace ID")
    # Is Stream Request
    is_stream: bool = Field(False, description="Is Stream Request")
    # Request path and method
    request_path: Optional[str] = Field(None, description="Request Path")
    request_url: Optional[str] = Field(None, description="Original Request URL")
    request_method: Optional[str] = Field(None, description="Request HTTP Method")
    # Upstream URL (full URL sent to provider)
    upstream_url: Optional[str] = Field(None, description="Upstream URL")
    # Protocol Conversion Fields (for debugging and analysis)
    # Client request protocol (openai/anthropic)
    request_protocol: Optional[str] = Field(None, description="Client Request Protocol")
    # Upstream supplier protocol (openai/anthropic)
    supplier_protocol: Optional[str] = Field(None, description="Upstream Supplier Protocol")
    # Converted request body (sent to upstream after protocol conversion)
    converted_request_body: Optional[dict[str, Any]] = Field(
        None, description="Converted Request Body (after protocol conversion)"
    )
    # Upstream response body (original response before protocol conversion)
    upstream_response_body: Optional[str] = Field(
        None, description="Upstream Response Body (before protocol conversion)"
    )


class RequestLogModel(RequestLogCreate):
    """Request Log Complete Model"""
    
    id: int = Field(..., description="Log ID")
    detail_available: bool = Field(
        True, description="Whether request detail data is still available"
    )
    
    model_config = ConfigDict(from_attributes=True)


class RequestLogSummary(BaseModel):
    """Request Log Summary Model (for list queries, no large fields)"""

    id: int = Field(..., description="Log ID")
    request_time: datetime = Field(..., description="Request Time")
    api_key_id: Optional[int] = Field(None, description="API Key ID")
    api_key_name: Optional[str] = Field(None, description="API Key Name")
    user_id: Optional[str] = Field(None, description="User ID")
    requested_model: Optional[str] = Field(None, description="Requested Model Name")
    target_model: Optional[str] = Field(None, description="Target Model Name")
    provider_id: Optional[int] = Field(None, description="Provider ID")
    provider_name: Optional[str] = Field(None, description="Provider Name")
    retry_count: int = Field(0, description="Retry Count")
    matched_provider_count: Optional[int] = None
    first_byte_delay_ms: Optional[int] = None
    total_time_ms: Optional[int] = None
    input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    total_cost: Optional[float] = None
    input_cost: Optional[float] = None
    output_cost: Optional[float] = None
    response_status: Optional[int] = None
    trace_id: Optional[str] = None
    is_stream: bool = False
    is_completed: bool = True
    retry_attempt_count: int = Field(0, description="Number of retry attempt logs")
    retry_attempts: list["RequestLogSummary"] = Field(
        default_factory=list, description="Failed provider attempts for this request"
    )

    model_config = ConfigDict(from_attributes=True)

    @field_validator("request_time", mode="after")
    @classmethod
    def _request_time_utc(cls, v: datetime) -> datetime:
        dt = ensure_utc(v)
        assert dt is not None
        return dt


class RequestLogResponse(RequestLogBase):
    """Request Log Response Model (List View)"""
    
    id: int = Field(..., description="Log ID")
    retry_count: int = Field(0, description="Retry Count")
    matched_provider_count: Optional[int] = None
    first_byte_delay_ms: Optional[int] = None
    total_time_ms: Optional[int] = None
    input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    total_cost: Optional[float] = None
    input_cost: Optional[float] = None
    output_cost: Optional[float] = None
    response_status: Optional[int] = None
    trace_id: Optional[str] = None
    is_stream: bool = False
    is_completed: bool = True
    retry_attempt_count: int = Field(0, description="Number of retry attempt logs")
    retry_attempts: list["RequestLogResponse"] = Field(
        default_factory=list, description="Failed provider attempts for this request"
    )
    
    model_config = ConfigDict(from_attributes=True)


class RequestLogDetailResponse(RequestLogModel):
    """Request Log Detail Response Model"""

    response_body: Optional[Any] = Field(None, description="Response Body (Auto-parsed JSON)")
    
    model_config = ConfigDict(from_attributes=True)


class RequestLogQuery(BaseModel):
    """Request Log Query Conditions"""

    # Time Range
    start_time: Optional[datetime] = Field(None, description="Start Time")
    end_time: Optional[datetime] = Field(None, description="End Time")
    # Relative time range preset (e.g. "24h"). Ignored when start_time is provided.
    timeline: Optional[str] = Field(
        None,
        pattern="^(1h|3h|6h|12h|24h|1w)$",
        description="Relative time range preset. Ignored when start_time is provided.",
    )
    # Model Filter
    requested_model: Optional[str] = Field(None, description="Requested Model (Fuzzy Match)")
    target_model: Optional[str] = Field(None, description="Target Model (Fuzzy Match)")
    # Provider Filter
    provider_id: Optional[int] = Field(None, description="Provider ID")
    # Status Code Filter
    status_min: Optional[int] = Field(None, description="Min Status Code")
    status_max: Optional[int] = Field(None, description="Max Status Code")
    # Error Filter
    has_error: Optional[bool] = Field(None, description="Has Error")
    # API Key Filter
    api_key_id: Optional[int] = Field(None, description="API Key ID")
    api_key_name: Optional[str] = Field(None, description="API Key Name")
    # User ID Filter
    user_id: Optional[str] = Field(None, description="User ID (Fuzzy Match)")
    # Retry Count Filter
    retry_count_min: Optional[int] = Field(None, description="Min Retry Count")
    retry_count_max: Optional[int] = Field(None, description="Max Retry Count")
    # Token Filter
    input_tokens_min: Optional[int] = Field(None, description="Min Input Tokens")
    input_tokens_max: Optional[int] = Field(None, description="Max Input Tokens")
    # Time Filter
    total_time_min: Optional[int] = Field(None, description="Min Total Time (ms)")
    total_time_max: Optional[int] = Field(None, description="Max Total Time (ms)")
    # Pagination
    page: int = Field(1, ge=1, description="Page Number")
    page_size: int = Field(20, ge=1, le=100, description="Items Per Page")
    # Sorting
    sort_by: str = Field("request_time", description="Sort Field")
    sort_order: str = Field("desc", pattern="^(asc|desc)$", description="Sort Order")
    # Is Completed Filter
    is_completed: Optional[bool] = Field(None, description="Is Completed")

    @field_validator("start_time", "end_time", mode="after")
    @classmethod
    def _query_time_utc(cls, v: Optional[datetime]) -> Optional[datetime]:
        return ensure_utc(v)


class LogCostStatsQuery(BaseModel):
    """Cost statistics query conditions"""

    # Time Range
    start_time: Optional[datetime] = Field(None, description="Start Time")
    end_time: Optional[datetime] = Field(None, description="End Time")
    # Relative time range preset (e.g. "24h"). Ignored when start_time is provided.
    timeline: Optional[str] = Field(
        None,
        pattern="^(1h|3h|6h|12h|24h|1w)$",
        description="Relative time range preset. Ignored when start_time is provided.",
    )
    # Core dimensions
    requested_model: Optional[str] = Field(None, description="Requested Model (Exact or fuzzy)")
    provider_id: Optional[int] = Field(None, description="Provider ID")
    api_key_id: Optional[int] = Field(None, description="API Key ID")
    api_key_name: Optional[str] = Field(None, description="API Key Name (Fuzzy Match)")
    user_id: Optional[str] = Field(None, description="User ID (Fuzzy Match)")
    # Bucket granularity: minute/hour/day
    bucket: str = Field("day", pattern="^(minute|hour|day)$", description="Trend bucket")
    # Minute bucket size (only used when bucket="minute")
    bucket_minutes: Optional[int] = Field(
        None,
        ge=1,
        le=1440,
        description="Minute bucket size (used when bucket=minute)",
    )
    # Timezone offset (minutes) applied to request_time before bucketing.
    # Example: UTC+8 => 480, UTC-8 => -480.
    tz_offset_minutes: int = Field(
        0,
        ge=-14 * 60,
        le=14 * 60,
        description="Timezone offset minutes for bucketing (UTC to local)",
    )
    # Group by dimension for model stats: request_model (default) or provider_model
    group_by: str = Field(
        "request_model",
        pattern="^(request_model|provider_model)$",
        description="Group by dimension for model stats",
    )

    @field_validator("start_time", "end_time", mode="after")
    @classmethod
    def _stats_time_utc(cls, v: Optional[datetime]) -> Optional[datetime]:
        return ensure_utc(v)


class LogCostSummary(BaseModel):
    """Aggregated cost summary"""

    request_count: int = 0
    success_count: int = 0
    failure_count: int = 0
    success_rate: float = 0.0
    total_cost: float = 0.0
    input_cost: float = 0.0
    output_cost: float = 0.0
    input_tokens: int = 0
    output_tokens: int = 0


class LogCostTrendPoint(BaseModel):
    """Cost trend point"""

    bucket: datetime
    request_count: int = 0
    total_cost: float = 0.0
    input_cost: float = 0.0
    output_cost: float = 0.0
    input_tokens: int = 0
    output_tokens: int = 0
    error_count: int = 0
    success_count: int = 0

    @field_validator("bucket", mode="after")
    @classmethod
    def _bucket_utc(cls, v: datetime) -> datetime:
        dt = ensure_utc(v)
        assert dt is not None
        return dt


class LogCostByModel(BaseModel):
    """Cost grouped by requested model"""

    requested_model: str
    request_count: int = 0
    total_cost: float = 0.0
    input_tokens: int = 0
    output_tokens: int = 0


class ModelCallStats(BaseModel):
    """Call health stats grouped by provider and actual model"""

    provider_name: str
    model_name: str
    request_count: int = 0
    success_count: int = 0
    failure_count: int = 0
    success_rate: float = 0.0
    avg_first_byte_time_ms: Optional[float] = None
    max_first_byte_time_ms: Optional[float] = None


class LogCostStatsResponse(BaseModel):
    """Cost stats response"""

    summary: LogCostSummary
    trend: list[LogCostTrendPoint]
    by_model: list[LogCostByModel]
    by_model_tokens: list[LogCostByModel] = Field(default_factory=list)
    model_call_stats: list[ModelCallStats] = Field(default_factory=list)


class ModelStats(BaseModel):
    """Aggregated model stats based on request logs"""

    requested_model: str
    avg_response_time_ms: Optional[float] = None
    avg_first_byte_time_ms: Optional[float] = None
    success_rate: float = 0.0
    failure_rate: float = 0.0


class ModelProviderStats(BaseModel):
    """Aggregated model/provider stats based on request logs"""

    requested_model: str
    target_model: str
    provider_name: str
    avg_first_byte_time_ms: Optional[float] = None
    avg_response_time_ms: Optional[float] = None
    success_rate: float = 0.0
    failure_rate: float = 0.0


class ApiKeyMonthlyCost(BaseModel):
    """API Key monthly cost summary"""

    api_key_id: int
    total_cost: float = 0.0
